docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+# Clear Containers Architecture
+
+* [Overview](#overview)
+    * [Hypervisor](#hypervisor)
+    * [Agent](#agent)
+    * [Runtime](#runtime)
+        * [create](#create)
+        * [start](#start)
+        * [exec](#exec)
+        * [kill](#kill)
+        * [delete](#delete)
+    * [Proxy](#proxy)
+    * [Shim](#shim)
+    * [Networking](#networking)
+* [Appendices](#appendices)
+    * [DAX](#dax)
+    * [Previous Releases](#previous-releases)
+        * [V2.0](#version-20)
+        * [V1.0](#version-10)
+
+## Overview
+
+This is an architectural overview of Clear Containers, based on the 3.0 release.
+
+The [Clear Containers runtime (cc-runtime)](https://github.com/clearcontainers/runtime)
+complies with the [OCI](https://github.com/opencontainers) specifications and thus
+works seamlessly with the [Docker Engine](https://www.docker.com/products/docker-engine)
+pluggable runtime architecture. In other words, one can transparently replace the
+[default Docker runtime (runc)](https://github.com/opencontainers/runc) with `cc-runtime`.
+
+![Runtime and virtcontainers](runtime-vc-relationship.png)
+![Docker and Clear Containers](docker-cc.png)
+
+`cc-runtime` creates a QEMU/KVM virtual machine for each container the Docker engine creates.
+
+The container process is then spawned by an [agent](https://github.com/clearcontainers/agent),
+running as a daemon on the guest operating system.
+Hyperstart opens 2 virtio serial interfaces (Control and I/O) on the guest, and QEMU exposes them
+as serial devices on the host. `cc-runtime` uses the control device for sending container
+management commands to the agent while the I/O serial device is used to pass I/O streams (`stdout`,
+`stderr`, `stdin`) between the guest and the Docker Engine.
+
+For any given container, both the init process and all potentially executed commands within that
+container, together with their related I/O streams, need to go through 2 virtio serial interfaces
+exported by QEMU. The [Clear Containers proxy (`cc-proxy`)](https://github.com/clearcontainers/proxy)
+multiplexes and demultiplexes those commands and streams for all container virtual machines.
+There is only one `cc-proxy` instance running per Clear Containers host.
+
+On the host, each container process is reaped by a Docker specific (`containerd-shim`) monitoring
```

Comment:
This is pretty docker specific.  We may want to make this more generic so it can cover !docker as well.

Reply:
done